fix num_assistant_tokens with heuristic schedule

jmamou opened the pull request:
# What does this PR do?

We have defined 2 different num_assistant_tokens_schedule values:
 - `heuristic`: When all _speculative_ tokens are correct, increase `num_assistant_tokens` by 2 else reduce by 1. `num_assistant_tokens` value is persistent over multiple generation calls with the same assistant model.
- `heuristic_transient`: Same as `"heuristic` but `num_assistant_tokens` is reset to its initial value after each generation call.

Fixes # (issue)
https://github.com/huggingface/transformers/pull/27979#issuecomment-1908153882

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

@gante @amyeroberts

jmamou committed:
fix heuristic num_assistant_tokens_schedule

gante commented on src/transformers/generation/configuration_utils.py:
```diff
@@ -246,8 +246,9 @@ class GenerationConfig(PushToHubMixin):
             Defines the schedule at which max assistant tokens shall be changed during inference.
-            - `"_heuristic_`: When all _speculative_ tokens are correct, increase `num_assistant_tokens` by 2 else
-              reduce by 1
+            - `"_heuristic_"`: When all _speculative_ tokens are correct, increase `num_assistant_tokens` by 2 else
```
```suggestion
            - `"heuristic"`: When all speculative tokens are correct, increase `num_assistant_tokens` by 2 else
```
(let's remove the `_`)

gante commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -225,7 +225,8 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
         # probably can be improved -- we want to balance the benefits of getting assistant tokens correct with the
         # cost of forecasting incorrect assistant tokens.
-        if self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic":
+        if self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic" or \
+            self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic_transient":
```
```suggestion
        if self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic" or self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic_transient":
```

Run `make fixup` in your terminal afterwards, and commit the result. Our CI will complain otherwise :)

gante reviewed (approved):
Thank you for the fix and the tests 💛

gante commented on src/transformers/generation/utils.py:
```diff
@@ -4507,6 +4507,8 @@ def assisted_decoding(
         if streamer is not None:
             streamer.end()
 
+        if candidate_generator.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic":
```
We need to check first that `candidate_generator.assistant_model` exists -- ngram speculation doesn't have this attribute, which is causing its tests to fail

jmamou replied:
sure!
also relevant for speculation with PLD

jmamou committed:
Update src/transformers/generation/configuration_utils.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

jmamou committed:
Update src/transformers/generation/candidate_generator.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

jmamou committed:
Update utils.py

check that candidate_generator.assistant_model exists since some some speculations (like ngram and PLD) don't have assistant_model attribute

ArthurZucker commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -225,7 +225,7 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
         # probably can be improved -- we want to balance the benefits of getting assistant tokens correct with the
         # cost of forecasting incorrect assistant tokens.
-        if self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic":
+        if self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic" or self.assistant_model.generation_config.num_assistant_tokens_schedule == "heuristic_transient":
```
```suggestion
        if self.assistant_model.generation_config.num_assistant_tokens_schedule in {"heuristic", "heuristic_transient"}:
```

ArthurZucker replied:
a nit and should be alright

ArthurZucker reviewed (commented):
Thanks for the PR!

jmamou committed:
Update src/transformers/generation/candidate_generator.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

amyeroberts commented on tests/generation/test_utils.py:
```diff
@@ -3454,3 +3454,49 @@ def prepare_inputs_for_generation(self, *args, foo=False, encoder_outputs=None,
+        }
+        model.generate(**inputs, **generation_kwargs)
+        # update_candidate_strategy is called only once and therefore, assistant_model.generation_config.num_assistant_tokens should be either 4 or 7
+        self.assertTrue(assistant_model.generation_config.num_assistant_tokens == 4 or assistant_model.generation_config.num_assistant_tokens == 7)
```
```suggestion
        self.assertTrue(assistant_model.generation_config.num_assistant_tokens in (4, 7))
```

amyeroberts reviewed (commented):
Thanks for adding this! 

For the code quality checks, you'll need to run `make fixup` and push the changes. 

From the docstring, it mentions behaviour about `"heuristic_transient" being reset, but I don't see logic relating to it in this diff. Does this already happen?

jmamou committed:
Update tests/generation/test_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jmamou committed:
make fixup

jmamou committed:
merge conflict

jmamou committed:
Merge branch 'fix_num_assistant_tokens' of https://github.com/jmamou/transformers into fix_num_assistant_tokens

jmamou committed:
fix docstring

jmamou commented:
> Thanks for adding this!
> 
> For the code quality checks, you'll need to run `make fixup` and push the changes.
> 
> From the docstring, it mentions behaviour about `"heuristic_transient" being reset, but I don't see logic relating to it in this diff. Does this already happen?

I just fixed docstring

amyeroberts commented:
@jmamou Let's try and make the CI green :) You'll need to resolve the quality checks by running `make fixup` and pushing the changes. 

For the other failing tests, you'll need to try rebasing on main.

jmamou committed:
make fixup

jmamou committed:
Merge branch 'huggingface:main' into fix_num_assistant_tokens

jmamou commented:
@amyeroberts 
tests_torch failed. 
Are you familiar with the error?

jmamou committed:
Merge branch 'huggingface:main' into fix_num_assistant_tokens

amyeroberts commented:
@jmamou Tbh, I'm not sure what the cause of these failures are. I would first suggest rebasing on main to make sure you have all of the most recent commits. This will trigger a re-run of the CI too.

jmamou committed:
Merge branch 'huggingface:main' into fix_num_assistant_tokens

jmamou committed:
Merge branch 'huggingface:main' into fix_num_assistant_tokens

jmamou commented:
@amyeroberts 
same test still fails 👎

gante commented:
@jmamou Our apologies, [this PR](https://github.com/huggingface/transformers/pull/29027) fixed it. Could you try rebasing again please? 🤗

jmamou committed:
Merge branch 'huggingface:main' into fix_num_assistant_tokens

gante commented:
@amyeroberts both CI failures seem unrelated to this PR, including `check_repository_consistency` 👀 do you have an idea of what might be causing it?

amyeroberts commented:
@jmamou Apologies for all of the current issues you've been experiencing with unrelated failures on this PR. 

The two current batches of failing tests should have been resolved with #29037, #29043

Could you try one (final 🤞) rebase to get this CI green 🟢 ?

jmamou committed:
Merge branch 'huggingface:main' into fix_num_assistant_tokens

jmamou commented:
@amyeroberts 
tests passed on CI 👍

amyeroberts commented:
@jmamou Thanks for this contribution and your patience with our misbehaving CI!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28759). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
